lib/ansible/module_utils/facts.py

```diff
@@ -915,6 +927,104 @@ def get_memory_facts(self):
+            product_version='Version',
+            system_vendor='Manufacturer'
+        )
+
+        SMB_TYPE_CHASSIS = dict(
+            form_factor='Chassis Type',
+        )
+
+        for (k, v) in SMB_TYPE_BIOS.items():
+            if smbios_bin is not None:
+                (rc, out, err) = module.run_command('%s -t SMB_TYPE_BIOS' % (smbios_bin))
+                if rc == 0:
+                    smbios_dict = dict((line.split(':')[0].strip(), line.split(':')[1].strip()) for line in out.splitlines() if ':' in line )
+                    thisvalue = smbios_dict[v]
+                    try:
+                        json.dumps(thisvalue)
+                    except UnicodeDecodeError:
+                        thisvalue = "NA"
+
+                    self.facts[k] = thisvalue
+                else:
+                    self.facts[k] = 'NA'
+            else:
+                self.facts[k] = 'NA'
+
+        for (k, v) in SMB_TYPE_SYSTEM.items():
+            if smbios_bin is not None:
+                (rc, out, err) = module.run_command('%s -t SMB_TYPE_SYSTEM' % (smbios_bin))
+                if rc == 0:
+                    smbios_dict = dict((line.split(':')[0].strip(), line.split(':')[1].strip()) for line in out.splitlines() if ':' in line )
+                    thisvalue = smbios_dict[v]
+                    try:
+                        json.dumps(thisvalue)
+                    except UnicodeDecodeError:
+                        thisvalue = "NA"
+
+                    self.facts[k] = thisvalue
+                else:
+                    self.facts[k] = 'NA'
+            else:
+                self.facts[k] = 'NA'
+
+        for (k, v) in SMB_TYPE_CHASSIS.items():
+            if smbios_bin is not None:
+                (rc, out, err) = module.run_command('%s -t SMB_TYPE_CHASSIS' % (smbios_bin))
+                if rc == 0:
+                    smbios_dict = dict((line.split(':')[0].strip(), line.split(':')[1].strip()) for line in out.splitlines() if ':' in line )
+                    thisvalue = smbios_dict[v]
+                    try:
+                        json.dumps(thisvalue)
+                    except UnicodeDecodeError:
+                        thisvalue = "NA"
+
+                    self.facts[k] = thisvalue
+                else:
+                    self.facts[k] = 'NA'
+            else:
+                self.facts[k] = 'NA'
+
+    def get_device_facts(self):
+        self.facts['devices'] = {}
+
+        DISK_FACTS = ["Product", "Revision", "Serial No", "Size", "Vendor",
+                      "Hard Errors", "Soft Errors", "Transport Errors", "Media Error",
+                      "Predictive Failure Analysis", "Illegal Request",
+                     ]
+
+        d = {}
+        rc, out, err = module.run_command("/usr/bin/kstat -C sderr")
+        if rc != 0:
+            return dict()
+
+        sd_instances = set(sorted([line.split(':')[1] for line in out.split('\n') if line.startswith('sderr')]))
```

Comment:
Is there a reason to sort this?  Since we're just putting it into a set (to uniquify the list I assume) it doesn't seem necessary to sort.  Since we're just iterating over sd_instances, a frozenset is more appropriate than a set.